### PR TITLE
Adding reload to datasourceSrv

### LINKED ISSUE #30728: DataSourceSrv is not updated after creating a new data source via HTTP API
**What happened**:
DataSourceSrv is not updated after creating a new data source via HTTP API call. And need to refresh the page to be able to get added data source from DataSourceSrv.get(). Until the page is refreshed or updating the added data source, I'm getting the following error `Datasource named [Data Source Name] was not found`.

**What you expected to happen**:
After creating a new data source via HTTP API call, I'm able to get the added data source via `DataSourceSrv.get()`

**How to reproduce it (as minimally and precisely as possible)**:
1. Add a new data source
2. Return back without updating it
3. Try to get the data source via `DataSourceSrv.get()`

**Anything else we need to know?**:
I use `backendSrv.get('/api/datasources')` for getting the list and for each item use `DataSourceSrv.get()` for making query.
Also on the attached screen, I use the default method for adding a data source and this problem is reproduced too.

**Environment**:
- Grafana version: v7.3.7
- Data source type & version:  redis-datasource
- OS Grafana is installed on: Docker
- User OS & Browser: Mac OS, Chrome 87.0.4280.141
- Grafana plugins: redis-app

data-source-srv-problem-1

### PR DESCRIPTION
**Pull request by tolzhabayev**
**What this PR does / why we need it**:
Currently, `DataSourceSrv` is not updated after creating a new data source via HTTP API call. Users would have to refresh the page to be able to get added data source from `DataSourceSrv.get()`. This PR introduced a `reload` method, that developers can use in order to make sure the `DataSourceSrv` is up to date. 

**Which issue(s) this PR fixes**:

Fixes #30728

### CODE REVIEW

File: public/app/features/plugins/datasource_srv.ts [hunk 1]
```diff
@@ -324,6 +325,11 @@ export class DatasourceSrv implements DataSourceService {
       };
     });
   }
+
+  async reload() {
+    const settings = await getBackendSrv().get('/api/frontend/settings');
+    getDatasourceSrv().init(settings.datasources, settings.defaultDatasource);
```
Comment (leventebalogh) [inline public/app/features/plugins/datasource_srv.ts:331]: **Question:** couldn't we just use `this.init(settings.datasources, settings.defaultDatasource)`?

As far as I understand this is the singleton instance that get's returned by `getDatasourceSrv()`, but can be that I am wrong.

File: public/app/features/plugins/tests/datasource_srv.test.ts [hunk 2]
```diff
@@ -294,5 +311,18 @@ describe('datasource_srv', () => {
...
+      getBackendSrvGetMock.mockReturnValueOnce({
+        datasources: {
+          ...dataSourceInit,
+        },
+        defaultDatasource: 'aaa',
+      });
+
+      await dataSourceSrv.reload();
+      expect(getBackendSrvGetMock).toHaveBeenCalledWith('/api/frontend/settings');
+      expect(getDataSourceSrvInitMock).toHaveBeenCalledWith(dataSourceInit, 'aaa');
```
Comment (leventebalogh) [inline public/app/features/plugins/tests/datasource_srv.test.ts:325]: 👏

**Review by leventebalogh — APPROVED** [review]
LGTM 👍

File: public/app/features/plugins/datasource_srv.ts (hunk 1 above)
Comment (tolzhabayev) [reply public/app/features/plugins/datasource_srv.ts]: @leventebalogh true, would be the same. Changed: https://github.com/grafana/grafana/pull/44217/commits/22ead3fee1dd032eff7808f71ad89887afb6bb5f

**Review by leventebalogh — APPROVED** [review]

**Review by mckn — APPROVED** [review]
LGTM!

**Review by ryantxu — APPROVED** [review]

File: public/app/features/plugins/datasource_srv.ts [hunk 3]
```diff
@@ -324,6 +325,11 @@ export class DatasourceSrv implements DataSourceService {
       };
     });
   }
+
+  async reload() {
+    const settings = await getBackendSrv().get('/api/frontend/settings');
+    this.init(settings.datasources, settings.defaultDatasource);
+  }
```
Comment (ryantxu) [inline public/app/features/plugins/datasource_srv.ts:332]: There is similar code in a redux action:
https://github.com/grafana/grafana/blob/main/public/app/features/datasources/state/actions.ts#L286

That one mutates config -- perhaps we should do the same here and then have the reduc action call `reload()`?

Comment (tolzhabayev) [reply public/app/features/plugins/datasource_srv.ts:335]: @ryantxu good call: https://github.com/grafana/grafana/pull/44217/commits/38b3adcfa3c6aeb4dcecdc281def2b27202f4a19

**Review by dprokop — APPROVED** [review]
LGTM. Doesn't it deserve a changelog mention?

**Review by mckn — APPROVED** [review]
LGTM! Thanks for all the great input @ryantxu

**Pull request merged**